### ISSUE TITLE
Deprecate or fix toPayload()?

### ISSUE THREAD

**Issue by karptonite** [ngrx/platform#132]
This is related to the discussion of the Action interface, [here](https://github.com/ngrx/platform/issues/128#issuecomment-316680916 ).
While there may be type safety issues with having action on the interface, not having payload leads to the odd implementation of toPayload, which both requires casting of action to `any` and breaks type safety on the return value:
```typescript
export function toPayload(action: Action): any {
  return (action as any).payload;
}
```
If  we want ngrx to be agnostic about how payloads are handled, and to help with type safety, the toPayload function seems to work against that. perhaps `toPayload()` should be deprecated? It will usually be preferable to use something like:
```typescript
.map( action: SpecificActionWithPayload => action.payload )
```

Alternatively, it might be possible to create an interface for ngrx
```typescript
export interface ActionWithPayload<T> extends Action {
  payload: T;
}
```
Then implement toPayload as follows:
```typescript
export function toPayload(action: ActionWithPayload<T>): T {
  return action.payload;
}
```
That should allow for type safety, assuming that the type of the action itself is previously known, maybe with something like:
```typescript
.map( action => action as SpecificActionWithPayload )
```
clunky, but until typescript can infer the action type via ofType, maybe an OK option?

**Comment by ovangle** [comment]
> until typescript can infer the action type via ofType

Typescript will *never* be able to infer the action type via `ofType` while `ofType` accepts a string constant. That was one of the reasons I made my proposal.

**Comment by karptonite** [comment]
@ovangle I'm not sure why you say that, although perhaps you understand typescript's inferences better than I do.

However, TypeScript CAN infer the type in a reducer, when using a switch statement to filter on the action type, which, as you say, is only a string. For example, for an action like this:

```typescript
export class LoadBegunAction implements Action {
	readonly type = LOAD_BEGUN;
	constructor( public payload: { item: ApiId } ) {}
}
```
Note the `readonly` and the use of a `const` for the type. TypeScript infers the type of the action in this switch statement:

```typescript
switch ( action.type ) {
	case essentialItemActions.LOAD_BEGUN:
		// Do stuff with action.payload, where the compiler has inferred the type of action
```
Obviously, `ofType` is a different situation, but it doesn't seem so different to me that I couldn't imagine the TypeScript devs figuring out a way to infer type in a function like `ofType` at some point.

**Comment by ovangle** [comment]
There is no type inference by the typescript compiler going on there, it's just javascript dynamic inspection performed by your editor of choice.

Consider the following:
```
import {Action} from '@ngrx/store';

const LOAD_BEGUN = 'load begun';
const LOAD_END = 'load end';

class LoadBegunAction implements Action {
  readonly type = LOAD_BEGUN;
  constructor(public payload: {item1: string}) {}
}

class LoadEndAction implements Action {
  readonly type = LOAD_END;
  constructor(public payload: {item2: string}) {}
}

function reducer(state, action: Action) {
  switch (action.type) {
    case LOAD_BEGUN:
      // Webstorm offers me the choices `item1` and `item2` here, because it knows I'm accessing
      // the payload of an object I've defined.
      // However the typescript compiler will emit no errors if I finish this with `item2`.
      action.payload.
  }
}
```

**Comment by ovangle** [comment]
You're looking to get something for nothing, the typescript compiler isn't magic, it's a library just like this one. It has no idea what you _intend_ to mean when you use the typestring in a case statement.

If you want stuff like this to be possible, you actually have to _expose type information that the compiler can use_. And adding a `PayloadAction<T>` interface doesn't help when you're using strings as proxies for the types. 

Typescript will be able to do what you want if you _pass round the types themselves_. See my (prematurely closed) issue for an example of how to accomplish this.

**Comment by karptonite** [comment]
> There is no type inference by the typescript compiler going on there, it's just javascript inspections performed by your editor of choice.

I believe you are mistaken. Your example didn't work work for two reasons: First, you didn't sufficiently restrict the type of your reducer. Second, you used PHPStorm, which does not appear to use the typescript compiler and other tools for type inference. I suggest you try VS Code if you want to see the types that TypeScript actually infers.

Consider the following example:
```typescript
class Foo1 {
	readonly type = 'foo1';
	public bar = 'bar';
}

class Foo2 {
	readonly type = 'foo2';
}

type Foos = Foo1 | Foo2;

function processFoo ( foo: Foos) {
	switch (foo.type) {
		case 'foo1':
			console.log(foo.bar);
	}
}
```
This compiles fine. But if you change the `readonly`'s to `public`, it doesn't compile at all, and errors with `Property 'bar' does not exist on type 'Foos'.
  Property 'bar' does not exist on type 'Foo2'.`. If this were just an issue of javascript inspections, the compiler would not fail to compile in the second case.

if the incoming type is sufficiently restricted, toPayload can user the compiler to infer the type of `payload`. I have a working example here: https://github.com/ngrx/platform/pull/157

**Comment by bfricka** [comment]
@karptonite, Quick niggle: JetBrains stuff _does_ use TypeScript service, but in older versions (or projects created in older versions), you had to specifically enable the TypeScript service (Settings -> TypeScript -> Use TypeScript Service)

**Comment by karptonite** [comment]
@bfricka Maybe the use Typescript service for some stuff, but they don't seem to use it for their type inferences, at least not exclusively. If they did, they wouldn't be having bugs like this one: https://youtrack.jetbrains.com/issue/WEB-27891. I have the correct setting to use Typescript services set.

**Comment by bfricka** [comment]
That's correct. They don't use the VS Code language protocol (there is an open issue for that), and they do a lot of their own processing. I wish they'd just go to use the language protocol :(

**Comment by ovangle** [comment]
I stand corrected on that point. I didn't think such an inference would be reliable enough to add to the compiler, but maybe I'm just unclear on the semantics of `readonly`.

**Comment by brandonroberts** [comment]
Closed via #266
